MSC1711: X.509 certificate verification for federation connections

richvdh opened the pull request:
[Rendered](https://github.com/matrix-org/matrix-doc/blob/rav/proposal/x509-for-federation/proposals/1711-x509-for-federation.md)

ara4n commented on proposals/1711-x509-for-federation.md:
```diff
@@ -0,0 +1,238 @@
+
+## Background: the failure of the Perspectives approach
+
+The Perspectives approach replaces the conventional heirarchy of trust provided
```
hierarchy.

ara4n commented on proposals/1711-x509-for-federation.md:
```diff
@@ -0,0 +1,238 @@
+it can be difficult to get a certificate for a `.onion` domain (again, Let's
+Encrypt do not support them).
+
+The reasons for requiring a signed certificate (or indeed, for using TLS at
```
i thought you really do want to use TLS, even when on Tor, unless you implicitly trust your Tor nodes and given that the exit traffic is unencrypted?

uhoreg replied:
If you're using Tor to visit a normal site, then yes, you definitely want to use TLS.  But if you're visiting a `.onion` site, then the traffic is all automatically encrypted (and verifiable) from client to server trough Tor, so there is less need for TLS over Tor.

jcgruenhage replied:
Maybe it would be a good idea make clear text targets configurable where TLS shouldn't be used, like .onion and .i2p, as well as IPs on e2ee overlay networks like hyperboria and yggdrasil.

richvdh replied:
not using TLS at all could be tricky, as it might mean the federation port needs to support both TLS and non-TLS connections. But certainly disabling cert checks for certain domains sounds plausible.

jcgruenhage replied:
In don't think that a single port would need to support TLS and non-TLS traffic for this, since hiddenservice.onion:8448 could be pointed towards localhost:8008 by Tor.

richvdh replied:
> In don't think that a single port would need to support TLS and non-TLS traffic for this, since hiddenservice.onion:8448 could be pointed towards localhost:8008 by Tor.

... except that anyone connecting to hiddenservice.onion:8448 will be expecting a TLS response, and in the general case, synapse does not expose the federation endpoints on port 8008. Obviously all this can be worked around, but it feels overcomplex compared to a configuration setting which simply says "skip the x509 checks for domains matching *.onion".

jcgruenhage replied:
I think this can be resolved by recommending that homeservers can be configured to skip x509 checks for certain IP ranges and domains.

richvdh replied:
Yup, I've tried to clarify this. Thanks.

ara4n commented:
This generally looks good to me, although it's a bit sad that people who have got used to self-signed certs magically working (like me on arasphere) will be forced to pull their LE lives together.

Is there any way to (optionally) fall back to tofu when you see a self-signed cert?  (asking somewhat rhetorically, given it feels it makes it too easy for an attacker to MITM new connections to servers via a self-signed cert)

jevolk commented:
related #1685

jcgruenhage commented on proposals/1711-x509-for-federation.md:
```diff
@@ -0,0 +1,238 @@
+
+It is not obvious to the author of this proposal that DANE provides any
+material advantages over the CA model. In particular it replaces the
+centralised trust of the CAs with the centralised trust of the DNS registries.
```
DANE has an advantage here:

> Since DNSSEC provides authenticated denial of existence (allows a resolver to validate that a certain domain name does not exist), DANE enables an incremental transition to verified, encrypted SMTP without any other external mechanisms, as described by RFC 7672. A DANE record indicates that the sender must use TLS.

This means we could continue to TOFU for servers that don't have a TLSA record set.

richvdh replied:
Yeah maybe. I don't want to end up in a situation where people have to deploy DNSSEC and DANE to get a secure deployment.

jcgruenhage replied:
Cloudflare and OVH offer "one click" DNSSEC, it's really just a box to tick. Apart from that, there is CoreDNS which makes hosting signed entries yourself a piece of cake too. Setting that up isn't harder than setting up the average LE client.

Aside of how it's not that hard, increasing DNSSEC adoption would be nice to see, afaict matrix.org hasn't checked that box on cloudflare yet.

jcgruenhage replied:
FTR: I've changed my opinion on this a lot. Having DNSSEC/DANE be the only option to get a secure deployment is not acceptable.

jcgruenhage commented on proposals/1711-x509-for-federation.md:
```diff
@@ -0,0 +1,238 @@
+
+<a id="f1"/>[1] It's *possible* to set up homeservers to use servers other than
+`matrix.org` as notaries, but we are not aware of any that are set up this
+way. [↩](#a1)
```
Not really important, but at least all my servers have more notaries set.

maxidor replied:
Mines too, and so do all those I managed as part of my professional services.

richvdh replied:
reworded accordingly.

uhoreg commented on proposals/1711-x509-for-federation.md:
```diff
@@ -0,0 +1,238 @@
+In short: the server should present a certificate for the matrix domain
+(`matrix.org` in the above example). This ensures that traffic cannot be
+intercepted by a MitM who can control the DNS response for the `SRV` record
+(perhaps via cache-poisoning or falsifying DNS responses).
```
If we also allow the certificate to be for a specially-formed subdomain of the Matrix domain (e.g. maybe something like `matrix--[anything].matrix.org`), then this could allow domains to delegate their Matrix hosting without handing over a certificate for their main domain.

uhoreg replied:
The idea here is that the domain owners can delegate an unused hostname to the hosters, which would be a real hostname so the hosters could easily get a, for example, Let's Encrypt certificate.  But the general form of the allowed certificate names would make it highly unlikely to be used for a real website (hence the `matrix--` prefix).

richvdh replied:
This was certainly an interesting idea, but we've dismissed it for a couple of reasons:

* There's a security trap for anybody who lets people sign up for subdomains (which is certainly not an uncommon business model): if you can register for `delegated--matrix.example.com`, you get to intercept all the matrix traffic for `example.com`.
* Generally it feels quite unintuitive and violates the principle of least surprise.
* The fact that we can't find any prior art for this sets off alarm bells too.

uhoreg commented on proposals/1711-x509-for-federation.md:
```diff
@@ -0,0 +1,238 @@
+direct requests to `server.example.com`.
+
+In this case, `server.example.com` would be required to present a valid
+certificate for `server.example.com`.
```
I think this has the potential of interacting badly with MSC1714.  If the certificate is for `server.example.com`, and if the `matrix.org` server is unreachable, and either https://matrix.org disappears completely, or they delete their `.well-known/matrix/server` file, then there is nothing linking `matrix.org` with `server.example.com`.  Then if `example.net` sees an event from `matrix.org` and tries to verify it by asking another server, they will be presented with a server key that's signed by a certificate for `server.example.com` with no information as to whether that certificate was valid at the time for signing `matrix.org`'s server key.

richvdh replied:
yes it does, though we're abandoning MSC1714 anyway.

richvdh commented:
Note for anyone reading this: we see #1708 as a pre-requisite. It is fair to say that #1708 has had a less than rapturous reception.

jcgruenhage commented on proposals/1711-x509-for-federation.md:
```diff
@@ -0,0 +1,238 @@
+It has long been the view of the `matrix.org` administrators that IP literals
+are only really suitable for internal testing. Those who wish to use them for
+that purpose could either disable certificate checks inside their network, or
+use their own CA to issue certificates.
```
Further up in the MSC falling back to perspectives is listed as a temporary option, we could extend that option for IP literals and .onion addresses until we have a clean solution there too.

jcgruenhage replied:
If we decide to do this though, we should definitely add more default notaries. There are several public homeservers that we should be able to trust enough to use for this, for example weho.st.

4nd3r replied:
> we should be able to trust 

untill you lose trust and then you will create The Process which defines The Standards to avoid losing trust again. congratulations, you just reinvented the wheel.

jcgruenhage replied:
@4nd3r what do you mean exactly?

richvdh replied:
he means that it's hard to pick homeservers to act as notaries, and you risk ending up back at square one with a set of entities who control the trust, like the CAs.

My current feeling is that falling back to perspectives brings more complication and security problems than it is worth. Supporting IP literals is a non-goal anyway; I am not going to massively complicate the protocol to do so.

jcgruenhage replied:
But the problem with CA trust is that any one of them can issue a rogue certificate, with a notaries approach you'd need to overtake the majority for an attack, which is a lot harder. Of course I wouldn't trust tchncs.de to verify my certificates alone, but if tchncs.de and disroot.org and weho.st and matrix.org and kamax.io and hacklab.fi all say that the certificate is the same, then I'd trust it. The barrier for trust is waaay lower on a perspectives approach than with CAs, and it's also not necessary that people agree on the notaries. I have set homeservers of people I physically met and exchanged keys with as my notaries.

richvdh replied:
so are you arguing that the whole MSC is misguided?

jcgruenhage replied:
No, my point was more that the comment of @4nd3r doesn't really apply because trust means something entirely different in a perspectives approach than what it means in a CA approach.

Going with what everyone else is doing (the CA system) is definitely better than the current perspectives model. A perspectives model *could* work better, but it has more potential to fail in weird ways and isn't as tested out in the wild as TLS is.

I would prefer it if servers aren't identified by their domain names though, but by their signing key instead. If that was the case, they could just use a self signed certificate that uses that same key and be done with it. Resolving of server_name -> connection details including signing key would need to be secured in some way though, perhaps by serving that via HTTPS from .well-known or DNSSEC. 

For now though, going with the CA system is still a large step in the right direction. Supporting DANE additionally would be a bonus :)

4nd3r replied:
> if tchncs.de and disroot.org and weho.st and matrix.org and kamax.io and hacklab.fi all say that the certificate is the same, then I'd trust it.

i wouldn't, because building scalable and secure Trust is hard. i would trust CAs more than some HS operator, who "has been around long enough" and deemed trustworthy by people whom I do not know. CA approach maybe broken depending what your threat model is, but at least it's widely used, monitored and audited. so pick your CA carefully :see_no_evil: and if your thread actor is capable of abusing CAs, then you have much bigger problems than MITM with rogue certificates.

> trust means something entirely different in a perspectives approach than what it means in a CA approach

lets agree to disagree.

richvdh replied:
sounds like we're all basically agreed that perspectives must die, whether or not we later replace the CAs with servers-being-identified-by-keys [I agree that sounds nice, btw, but a way off yet.], so I'm marking this thread resolved.

richvdh commented on proposals/1711-x509-for-federation.md:
```diff
@@ -0,0 +1,238 @@
+In short: the server should present a certificate for the matrix domain
+(`matrix.org` in the above example). This ensures that traffic cannot be
+intercepted by a MitM who can control the DNS response for the `SRV` record
+(perhaps via cache-poisoning or falsifying DNS responses).
```
> we should accept a certificate with the hostname, 

This appears to be in conflict with what is specified for XMPP [1] and by RFC6125. Generally I'm not in favour of suddenly introducing different behaviour in the presence of DNSSEC.

1: https://tools.ietf.org/html/rfc6120#section-13.7.2.1
2: https://tools.ietf.org/html/rfc6125#section-6.2.1

maxidor replied:
RFC6125 defines out of scope the following:

> Keys or certificates employed outside the context of PKIX-based systems.
>
>    Some deployed application technologies use a web of trust model
>    based on or similar to OpenPGP [OPENPGP], or use self-signed
>    certificates, or are deployed on networks that are not directly
>    connected to the public Internet and therefore cannot depend on
>    Certificate Revocation Lists (CRLs) or the Online Certificate
>    Status Protocol [OCSP] to check CA-issued certificates.

Self-signed certificates are already accepted, while the issues you outlined in your proposal really refer to the Perspective approach. So thing like DNSSEC and DANE would work if the final certificate has the target hostname instead of the original hostname and not break the RFC.

RFC6120 is about XMPP, but we are not XMPP.

richvdh commented:
right, I have updated this, and moved out the MSC1708 section because really that belongs to MSC1708. Everyone seems to be basically on board with the principle of this MSC, so I'm going to propose a FCP.

@mscbot fcp merge

jcgruenhage commented on proposals/1711-x509-for-federation.md:
```diff
@@ -58,24 +58,31 @@ We propose that Matrix homeservers should be required to present valid TLS
+certificate checks for a whitelist of domains/netmasks. This would be useful
+for `.onion` domains (where a certificate is hard to obtain, and where server
+verification is provided at the network level), as well as for testing with IP
+literals.
```
My comment about disabling it for certain IP ranges wasn't about testing, but about e2ee networks, where a hash of the public key of a host is encoded in the address, like yggdrasil and cjdns.

richvdh replied:
ok. Can you suggest better wording?

jcgruenhage replied:
```md
It would also be useful for administrators to be able to disable the
certificate checks for a whitelist of domains/netmasks. This would be useful
for testing and networks that provide server verification themselves, 
like `.onion` domains on Tor, `.i2p` domains on I2P, 
`200::/7` IPs on yggdrasil and `fc00::/8` IPs on cjdns. 
```

This might be too much examples, feel free to cut that down to one domain and one IP example if you want that section to be shorter.

jcgruenhage replied:
Happy with the new revision! :tada:

ara4n commented:
/me wonders if there was any merit to his comment:

> Is there any way to (optionally) fall back to tofu when you see a self-signed cert? (asking somewhat rhetorically, given it feels it makes it too easy for an attacker to MITM new connections to servers via a self-signed cert)

ara4n commented:
answer: the problem is that even if you publish a self-signed cert, it needs the *other* servers in the rooms to agree to do tofu. if they don't, then it all fails. so this would effectively need all servers to do tofu, which unless you know up front all the servers who connect to you (i.e. private federation), means everyone has to do tofu, at which point it's not optional any more.

hawkowl commented on proposals/1711-x509-for-federation.md:
```diff
@@ -0,0 +1,237 @@
+most reverse-proxies will reject requests where the SNI and Host headers do not
+match. [↩](#a2)
+
+<a id="f3"/>[3] Let's Encrypt will issue ACME challenges via port 80 or DNS
```
This is incorrect -- as per discussions in meetings, Synapse being able to respond to HTTP-01 is fine, either through an authbind to port 80, or nginx/etc doing vhost forwarding.

richvdh replied:
sorry, which bit is incorrect?

richvdh replied:
oh, the next line. thanks github for making that so confusing.

mherrb commented:
when using an SRV DNS record to point to, let's say example.com, does it mean that 'example.com' needs to be include in the list of subjectAltNames in the certificate ?
(the tetaneutral.net domain is setup with a SRV record, but fails the certificate check on https://matrix.org/federationtester/ - I'm wondering if the missing tetaneutral.net domain is the cause.

jcgruenhage commented:
@mherrb You will want to either have your server name in your DNS SANs, or have a .well-known record pointing to the other server. https://github.com/matrix-org/synapse/blob/master/docs/MSC1711_certificates_FAQ.md has more info on what you can do.

mherrb commented:
Thanks.

turt2live commented:
This is merged via https://github.com/matrix-org/matrix-doc/pull/1830
